common/src/main/java/org/openshift/ping/common/stream/TokenStreamProvider.java

```diff
@@ -0,0 +1,101 @@
+package org.openshift.ping.common.stream;
+
+import static org.openshift.ping.common.Utils.openFile;
+
+import java.io.FileNotFoundException;
+
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URLConnection;
+import java.security.KeyStore;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+
+/**
+ * @author <a href="mailto:rmartine@redhat.com">Ricardo Martinelli</a>
+ */
+public class TokenStreamProvider extends BaseStreamProvider {
+	
+	private static final Logger log = Logger.getLogger(TokenStreamProvider.class.getName());
+
+	private String token;
+	
+    private final SSLSocketFactory factory;
+
+	public TokenStreamProvider(String token, String caCertFile) throws Exception {
+		this.token = token;
+        TrustManager[] trustManagers = configureCaCert(caCertFile);
+        SSLContext context = SSLContext.getInstance("TLS");
+        context.init(null, trustManagers, null);
+        factory = context.getSocketFactory();
+	}
+
+	@Override
+	public InputStream openStream(String url, Map<String, String> headers, int connectTimeout, int readTimeout)
+			throws IOException {
+        URLConnection connection = openConnection(url, headers, connectTimeout, readTimeout);
+
+        if (connection instanceof HttpsURLConnection) {
+            HttpsURLConnection httpsConnection = HttpsURLConnection.class.cast(connection);
+            //httpsConnection.setHostnameVerifier(InsecureStreamProvider.INSECURE_HOSTNAME_VERIFIER);
+            httpsConnection.setSSLSocketFactory(factory);
+            if (log.isLoggable(Level.FINE)) {
+                log.fine(String.format("Using HttpsURLConnection with SSLSocketFactory [%s] for url [%s].", factory, url));
+            }
+        } else {
+            if (log.isLoggable(Level.FINE)) {
+                log.fine(String.format("Using URLConnection for url [%s].", url));
+            }
+        }
+        
+		if (token != null) {
+            // curl -k -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
+            // https://172.30.0.2:443/api/v1/namespaces/dward/pods?labelSelector=application%3Deap-app
+            headers.put("Authorization", "Bearer " + token);
```

Comment:
I'm wondering whether we should send the bearer token when dealing with a non-TLS connection. Perhaps we should throw an error? The rationale is that if someone somehow points us to a MITM server through HTTP, we'll end up sending the token to the MITM. 
But, of course, not sending the bearer token would prevent us from ever using an unsecure OpenShift API server (during development, for example). Please ask others, what they think - or we can talk about it at the meeting tomorrow.